src/main/java/org/variantsync/diffdetective/variationtree/VariationTreeNode.java

```diff
@@ -0,0 +1,364 @@
+package org.variantsync.diffdetective.variationtree;
+
+import org.prop4j.Node;
+import org.variantsync.diffdetective.diff.DiffLineNumber;
+import org.variantsync.diffdetective.diff.difftree.DiffNode; // For Javdoc
+import org.variantsync.diffdetective.diff.difftree.DiffTree; // For Javdoc
+import org.variantsync.diffdetective.diff.difftree.DiffType;
+import org.variantsync.diffdetective.diff.difftree.NodeType;
+import org.variantsync.diffdetective.util.Assert;
+import org.variantsync.diffdetective.util.fide.FixTrueFalse;
+
+import java.util.*;
+
+/**
+ * A single node in a variation tree.
+ *
+ * <p>A variation tree is a representation of source code in a software product line. Each node in
+ * such a tree is either an {@link isArtifact artifact} or a {@link isAnnotation mapping node}.
+ * Artifacts represent source code which may be reused in multiple variations. In contrast, mappings
+ * do not contain source code of a specific variant. They store information about which variants
+ * contain the artifacts it has as {@link getChildren children} in the form of a
+ * {@link getFeatureMapping feature mapping}.
+ *
+ * <p>See definition 2.2 of <a href="https://doi.org/10.1145/3540250.3549108">our ESEC/FSE'22
+ * paper</a> for a mathematical formalization or variation trees. In contrast to this formalization,
+ * this implementation optimizes for variation trees obtained from source code annotated with the C
+ * preprocessor syntax. This mostly means that there is the additional {@link NodeType#ELIF} node
+ * type.
+ *
+ * <p>This class contains references to all of its children and its parent so all connected nodes of
+ * a variation tree can be reached through each node of the variation tree. Nevertheless, most of
+ * the time only node itself or its subtree (the reflexive hull of {@link getChildren}) is meant
+ * when when referencing a {@code VariationTreeNode}. Use {@link VariationTree} to unambiguously
+ * refer to a whole variation tree.
+ *
+ * <p>If possible, algorithms should be using {@link VariationNode} instead of this concrete
+ * implementation. This allows the usage of a projection of a {@link DiffNode} instead of a concrete
+ * variation node.
+ *
+ * <p>To compare two variation trees use {@link DiffTree} which uses the aforementioned
+ * {@link DiffNode}s.
+ *
+ * @see VariationTree
+ * @see DiffTree
+ * @see DiffNode
+ * @author Benjamin Moosherr
+ */
+public class VariationTreeNode extends VariationNode<VariationTreeNode> {
+    private static final short ID_OFFSET = 3;
```

Comment:
Maybe we should store this somewhere else because currently this is duplicated. The ID_OFFSET depends on the number of values of DiffType and NodeType (it says how many bits are necessary to store each one of them). What about putting this into NodeType?

Reply:
Good idea! I added some comments below.